balancer/rls/internal/picker.go

```diff
@@ -0,0 +1,394 @@
+//
+// Caller must hold a write-lock on p.lb.cacheMu.
+func (p *rlsPicker) sendRouteLookupRequest(cacheKey cacheKey, bs *backoffState, reqKeys map[string]string, reason rlspb.RouteLookupRequest_Reason, staleHeaders string) bool {
+	if p.lb.pendingMap[cacheKey] != nil {
+		return false
+	}
+
+	p.lb.pendingMap[cacheKey] = bs
+	throttled := p.ctrlCh.lookup(reqKeys, reason, staleHeaders, func(targets []string, headerData string, err error) {
+		p.handleRouteLookupResponse(cacheKey, targets, headerData, err)
+	})
+	if throttled {
+		delete(p.lb.pendingMap, cacheKey)
+	}
+	return throttled
+}
+
+// pickIfFeasible determines if a pick can be delegated to child policy based on
+// its connectivity state.
+// - If state is CONNECTING, the pick is to be queued
+// - If state is IDLE, the child policy is instructed to exit idle, and the pick
+//   is to be queued
+// - If state is READY, pick it delegated to the child policy's picker
+func (p *rlsPicker) pickIfFeasible(cpw *childPolicyWrapper, info balancer.PickInfo) (bool, balancer.PickResult, error) {
+	state := (*balancer.State)(atomic.LoadPointer(&cpw.state))
+	switch state.ConnectivityState {
+	case connectivity.Connecting:
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	case connectivity.Idle:
+		p.bg.ExitIdleOne(cpw.target)
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	case connectivity.Ready:
+		r, e := state.Picker.Pick(info)
+		return true, r, e
+	}
+	return false, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+}
+
+// handleRouteLookupResponse is the callback invoked by the control channel upon
+// receipt of an RLS response. Modifies the data cache and pending requests map
+// and sends a new picker.
+//
+// Acquires the write-lock on the cache. Caller must not hold p.lb.cacheMu.
+func (p *rlsPicker) handleRouteLookupResponse(cacheKey cacheKey, targets []string, headerData string, err error) {
+	p.logger.Infof("Received RLS response for key %+v with targets %+v, headerData %q, err: %v", cacheKey, targets, headerData, err)
+
+	p.lb.cacheMu.Lock()
+	defer func() {
+		// Pending request map entry is unconditionally deleted since the request is
+		// no longer pending.
+		p.logger.Infof("Removing pending request entry for key %+v", cacheKey)
+		delete(p.lb.pendingMap, cacheKey)
+		p.lb.sendNewPickerLocked()
+		p.lb.cacheMu.Unlock()
+	}()
+
+	// Lookup the data cache entry or create a new one.
+	dcEntry := p.lb.dataCache.getEntry(cacheKey)
+	if dcEntry == nil {
+		dcEntry = &cacheEntry{}
+		if _, ok := p.lb.dataCache.addEntry(cacheKey, dcEntry); !ok {
+			// This is a very unlikely case where we are unable to add a
+			// data cache entry. Log and leave.
+			p.logger.Warningf("Failed to add data cache entry for %+v", cacheKey)
+			return
+		}
+	}
+
+	// For failed requests, the data cache entry is modified as follows:
+	// - status is set to error returned from the control channel
```

Comment:
Is this right? What if it was a proactive lookup and the existing entry is still valid?  We want to actively invalidate it in the event of an error?

Reply:
Hmm ... there is no mention of any special treatment for this case (as opposed to regular RLS failure). I have posted a question on the RLS chat room. Will see what Mark says.

Reply:
I added a comment to clarify. Basically the code as written is doing the correct thing, which is to leave the targets and the expiration time on the old entry unchanged. So, the entry will be used until it expires, and since we start the backoff here, a new picker will be sent when the backoff expires.